--- gcc/varasm.c	2023-06-27 09:57:46.814372325 +0300
+++ gcc/varasm_patched.c	2023-06-27 10:00:16.027279107 +0300
@@ -1193,15 +1193,11 @@
 
   /* Tell assembler to move to target machine's alignment for functions.  */
   align = floor_log2 (FUNCTION_BOUNDARY / BITS_PER_UNIT);
-  if (align > 0)
-    {
-      ASM_OUTPUT_ALIGN (asm_out_file, align);
-    }
 
   /* Handle a user-specified function alignment.
      Note that we still need to align to FUNCTION_BOUNDARY, as above,
      because ASM_OUTPUT_MAX_SKIP_ALIGN might not do any alignment at all.  */
-  if (align_functions_log > align)
+  if ((align_functions_log > align) || (align > align_functions_log))
     {
 #ifdef ASM_OUTPUT_MAX_SKIP_ALIGN
       ASM_OUTPUT_MAX_SKIP_ALIGN (asm_out_file,
@@ -1209,6 +1205,11 @@
 #else
       ASM_OUTPUT_ALIGN (asm_out_file, align_functions_log);
 #endif
+    } else 
+  /* Tell assembler to move to target machine's alignment for functions.  */
+  if (align > 0)
+    {
+      ASM_OUTPUT_ALIGN (asm_out_file, align);
     }
 
 #ifdef ASM_OUTPUT_FUNCTION_PREFIX
